rfcs/0045-zkapp-balance-data-in-archive.md

```diff
@@ -0,0 +1,292 @@
+## Summary
+[summary]: #summary
+
+For zkApps, balances and account creation fees of the affected parties
+are not being stored in the database. The data sent from the daemon to
+the archive process does not contain such information. The archive
+database schema does not have a place to store all such information, were
+it provided.
+
+Add additional information to the RPC data sent from the daemon, so
+that account creation information and balances are provided for each
+block. Update the schema to allow storing the information.
+
+Currently, balance and nonce information is stored for each
+transaction, which is more fine-grained than needed. Instead, store
+complete account information, including balances and nonces, for each
+account with a transaction in a block, just once for that block. That
+should simplify both the code to store that information, and queries
+for that information, as done by the Rosetta implementation.
+
+We also wish to store account index information (ledger leaf order),
+to allow predicting block winners from archive database information.
+
+## Motivation
+[motivation]: #motivation
+
+The archive processor has no code to store balance and account
+creation fee information for zkApp transactions. The replayer needs
+such information in the archive database to verify balances and
+nonces, and Rosetta needs it to respond to account queries.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Changes to the archive database schema
+
+Balances are no longer per-transaction, but per-block.
+
+Table `blocks_internal_commands`, remove the column:
+
+  `receiver_account_creation_fee_paid`
+  `receiver_balance`
+
+Table `blocks_user_commands`, remove the columns:
+
+  `fee_payer_account_creation_fee_paid`
+  `receiver_account_creation_fee_paid`
+  `fee_payer_balance`
+  `source_balance`
+  `receiver_balance`
+
+where the array contains account creation fees for each of the `other_parties`.
+While the array is not nullable, array elements may be `NULL`.
+
+Rename the existing `zkapp_account` to `zkapp_precondition_accounts`, and in the
+table `zkapp_predicate`, rename `account_id` to `precondition_account_id`, and
+modify the foreign key reference accordingly.
+
+Add new table `zkapp_accounts`:
+```
+  app_state           int     NOT NULL  REFERENCES zkapp_states
+  verification_key    int     NOT NULL  REFERENCES zkapp_verification_keys(id)
+  zkapp_version       bigint  NOT NULL
+  sequence_state      int     NOT NULL  REFERENCES zkapp_sequence_states(id)
+  last_sequence_slot  bigint  NOT NULL
+  proved_state        bool    NOT NULL
+```
+
+The table `balances` is replaced by a new table `accounts_accessed`, with columns:
```

Comment:
Can you add `ledger_index int NOT NULL` to the account? We need this info for block production / VRF calculation, and to faithfully reconstruct the ledger, so it would be nice to add it.

Reply:
For a given public key, the ledger index remains the same, regardless of other changes to the associated account, so I think it's better if the index is stored just once in the `public_keys` table, rather than multiple times whenever the account data changes.

Reply:
This isn't necessarily true: if there's a short-range fork / re-org, then we may find the public key appear at a different index between the forks based on the order of transactions, so we need to be careful to identify the correct index.

Reply:
OK, done.